pkg/provider/vault/vault.go

```diff
@@ -805,41 +839,57 @@ func getCertFromConfigMap(v *client) ([]byte, error) {
 	return []byte(val), nil
 }
 
+/*
+setAuth gets a new token using the configured mechanism
+if there's already a valid token, does nothing
+*/
 func (v *client) setAuth(ctx context.Context, cfg *vault.Config) error {
-	tokenExists, err := setSecretKeyToken(ctx, v)
+	tokenExists, err := checkToken(ctx, v)
+	if tokenExists {
+		v.log.V(1).Info("Re-using existing token")
+		return err
+	}
+
```

Comment:
If the token is expired, we continue using the same cached client. We just request a new token for it. I assume the Vault client will clean up the old one when `requestTokenWithXXXAuth` is called.

Good point on the ResourceVersion changing. I actually think anytime a store is deleted or updated, there's no mechanism to remove the now-invalid item from the cache.

Any ideas on where to put clean-up code? Are there any methods that are automatically called when a SecretStore changes or is deleted? Lacking that, maybe this could be a fixed-size LRU cache or something? That would obviously complicate things significantly.

Reply:
I addressed the update case in the meantime, though.

Reply:
> Any ideas on where to put clean-up code?

[This](https://pkg.go.dev/github.com/hashicorp/golang-lru#section-readme) looks like a good replacement for the `map[clientCacheKey]clientCacheValue`, WDYT?

One thing that may be missing is when we delete an entry from the cache due to a ResourceVersion change we should cleanup/revoke the old token `if v.store.Auth.TokenSecretRef == nil`.

Reply:
Yeah, that should work if you don't mind pulling in a new dependency. Would probably be a little cleaner if we could detect when a store has been deleted, but from what I can see, that would require more extensive code changes. This will at least prevent unbounded growth of the cache.

Any thoughts on how large the cache should be? Does it need to be user-configurable?

Reply:
I'd make it user-configurable as it depends on a couple of user-specific factors:
* how much memory am i willing to allocate?
* how many secret stores a user has, how often they are updated
* external secret refresh interval
* authentication method / token lease time 

Maybe start with a size of 100 and go from there? :thinking:  :shrug:

I guess we would need more in-depth metrics to give the user better visibility - but that's out of scope for now i believe. Especially the calls that are being made towards vault would be of great interest.